python/pyspark/sql/functions.py

```diff
@@ -12451,37 +12451,132 @@ def inline(col: "ColumnOrName") -> Column:
+
     .. versionadded:: 3.4.0
 
     Parameters
     ----------
     col : :class:`~pyspark.sql.Column` or str
-        input column of values to explode.
+        Input column of values to explode.
 
     Returns
     -------
     :class:`~pyspark.sql.Column`
-        generator expression with the inline exploded result.
+        Generator expression with the inline exploded result.
 
     See Also
     --------
-    :meth:`explode`
-
-    Notes
-    -----
-    Supports Spark Connect.
+    :meth:`pyspark.functions.explode`
+    :meth:`pyspark.functions.inline_outer`
 
     Examples
     --------
+    Example 1: Using inline with a single struct array column
+
+    >>> import pyspark.sql.functions as sf
+    >>> from pyspark.sql import Row
+    >>> df = spark.createDataFrame([Row(structlist=[Row(a=1, b=2), Row(a=3, b=4)])])
+    >>> df.select(sf.inline(df.structlist)).show()
+    +---+---+
+    |  a|  b|
+    +---+---+
+    |  1|  2|
+    |  3|  4|
+    +---+---+
+    
+    Example 2: Using inline with a column name
+
+    >>> import pyspark.sql.functions as sf
     >>> from pyspark.sql import Row
     >>> df = spark.createDataFrame([Row(structlist=[Row(a=1, b=2), Row(a=3, b=4)])])
-    >>> df.select(inline(df.structlist)).show()
+    >>> df.select(sf.inline("structlist")).show()
     +---+---+
     |  a|  b|
     +---+---+
     |  1|  2|
     |  3|  4|
     +---+---+
+
+    Example 3: Using inline with an alias
+
+    >>> import pyspark.sql.functions as sf
+    >>> from pyspark.sql import Row
+    >>> df = spark.createDataFrame([Row(structlist=[Row(a=1, b=2), Row(a=3, b=4)])])
+    >>> df.select(sf.inline("structlist").alias("c1", "c2")).show()
+    +---+---+
+    | c1| c2|
+    +---+---+
+    |  1|  2|
+    |  3|  4|
+    +---+---+
+
+    Example 4: Using inline with multiple struct array columns
+
+    >>> import pyspark.sql.functions as sf
+    >>> from pyspark.sql import Row
+    >>> df = spark.createDataFrame([
+    ...     Row(structlist1=[Row(a=1, b=2), Row(a=3, b=4)],
+    ...         structlist2=[Row(c=5, d=6), Row(c=7, d=8)])
+    ... ])
+    >>> df.select(sf.inline("structlist1"), "structlist2") \
+    ...     .select("a", "b", sf.inline("structlist2")).show()
+    +---+---+---+---+
+    |  a|  b|  c|  d|
+    +---+---+---+---+
+    |  1|  2|  5|  6|
+    |  1|  2|  7|  8|
+    |  3|  4|  5|  6|
+    |  3|  4|  7|  8|
+    +---+---+---+---+
+
+    Example 5: Using inline with a nested struct array column
+
+    >>> import pyspark.sql.functions as sf
+    >>> from pyspark.sql import Row
+    >>> df = spark.createDataFrame([
+    ...     Row(structlist=[Row(a=1, b=2, nested=[Row(c=3, d=4), Row(c=5, d=6)])])
+    ... ])
+    >>> df.select(sf.inline("structlist")).show()
+    +---+---+----------------+
+    |  a|  b|          nested|
+    +---+---+----------------+
+    |  1|  2|[{3, 4}, {5, 6}]|
+    +---+---+----------------+
+    >>> df.select(sf.inline("structlist")).select(sf.inline("nested")).show()
```

Comment:
cc @zhengruifeng @HyukjinKwon Interestingly, if I directly use `df.select(sf.inline("structlist.nested"))` It throws an exception:
`
pyspark.errors.exceptions.captured.AnalysisException: [DATATYPE_MISMATCH.UNEXPECTED_INPUT_TYPE] Cannot resolve "inline(structlist.nested)" due to data type mismatch: Parameter 1 requires the "ARRAY<STRUCT>" type, however "structlist.nested" has the type "ARRAY<ARRAY<STRUCT<c: BIGINT, d: BIGINT>>>".;
`
Is this expected? Why is the schema of "nested" inferred as array<array<struct>>?

Reply:
```
In [14]: df.select("structlist").printSchema()
root
 |-- structlist: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- a: long (nullable = true)
 |    |    |-- b: long (nullable = true)
 |    |    |-- nested: array (nullable = true)
 |    |    |    |-- element: struct (containsNull = true)
 |    |    |    |    |-- c: long (nullable = true)
 |    |    |    |    |-- d: long (nullable = true)


In [15]: df.select("structlist.nested").printSchema()
root
 |-- nested: array (nullable = true)
 |    |-- element: array (containsNull = true)
 |    |    |-- element: struct (containsNull = true)
 |    |    |    |-- c: long (nullable = true)
 |    |    |    |-- d: long (nullable = true)


In [16]: df.select(df.structlist.nested).printSchema()
root
 |-- structlist.nested: array (nullable = true)
 |    |-- element: array (containsNull = true)
 |    |    |-- element: struct (containsNull = true)
 |    |    |    |-- c: long (nullable = true)
 |    |    |    |-- d: long (nullable = true)
```

I guess there is something wrong in column resolution?

Reply:
```
In [25]: df = spark.sql("SELECT ARRAY(STRUCT(a,b,c)) AS x FROM VALUES (1, 1.1, 'a'), (2, 2.2, 'b'), (4, 4.4, 'c') AS TAB(a, b,
    ...: c)")

In [26]: df
Out[26]: DataFrame[x: array<struct<a:int,b:decimal(2,1),c:string>>]

In [27]: df.createOrReplaceTempView("v")

In [28]: df.printSchema()
root
 |-- x: array (nullable = false)
 |    |-- element: struct (containsNull = false)
 |    |    |-- a: integer (nullable = false)
 |    |    |-- b: decimal(2,1) (nullable = false)
 |    |    |-- c: string (nullable = false)


In [29]: spark.sql("SELECT x.a FROM v").printSchema()
root
 |-- a: array (nullable = false)
 |    |-- element: integer (containsNull = false)

```

@allisonwang-db I think this is at least consistent with SQL, so should be fine.

Reply:
```
>>> df.select("structlist.nested").show()
+------------------+
|            nested|
+------------------+
|[[{3, 4}, {5, 6}]]|
+------------------+
```

it fails because `structlist.nested` is an array of an array of structs.


```
>>> df.select(sf.inline("structlist")).show()
+---+---+----------------+
|  a|  b|          nested|
+---+---+----------------+
|  1|  2|[{3, 4}, {5, 6}]|
+---+---+----------------+
>>> df.select(sf.inline("structlist")).select(sf.inline("nested")).show()
+---+---+
|  c|  d|
+---+---+
|  3|  4|
|  5|  6|
+---+---+
```

works because `nested` is an array of structs (?)

Reply:
yes, because both `structlist` and `nested` are `array<struct>`, but ``structlist.nested` is `array<array<<struct>>` (this is a SQL feature)

Reply:
I see. It's just strange to me that the type of `structlist.nested` is array<array<struct>>.